arch/arm/soc/nordic_nrf/nrf51/soc.c

```diff
@@ -25,52 +25,17 @@ extern void _NmiInit(void);
 #define NMI_INIT()
 #endif
 
-#include "nrf.h"
-
-#define __SYSTEM_CLOCK (16000000UL)
-
-static bool ftpan_26(void);
-static bool ftpan_59(void);
-
-uint32_t SystemCoreClock __used = __SYSTEM_CLOCK;
+#include <system_nrf51.h>
 
 static int nordicsemi_nrf51_init(struct device *arg)
 {
 	u32_t key;
 
 	ARG_UNUSED(arg);
 
-	/* Note:
-	 * Magic numbers below are obtained by reading the registers
-	 * when the SoC was running the SAM-BA bootloader
-	 * (with reserved bits set to 0).
-	 */
-
 	key = irq_lock();
 
-	/* Prepare the peripherals for use as indicated by the PAN 26 "System:
-	 * Manual setup is required to enable the use of peripherals" found at
-	 * Product Anomaly document for your device found at
-	 * https://www.nordicsemi.com/. The side effect of executing these
-	 * instructions in the devices that do not need it is that the new
-	 * peripherals in the second generation devices (LPCOMP for example)
-	 * will not be available.
-	 */
-	if (ftpan_26()) {
-		*(volatile u32_t *)0x40000504 = 0xC007FFDF;
-		*(volatile u32_t *)0x40006C18 = 0x00008000;
-	}
-
-	/* Disable PROTENSET registers under debug, as indicated by PAN 59
-	 * "MPU: Reset value of DISABLEINDEBUG register is incorrect" found
-	 * at Product Anomaly document for your device found at
-	 * https://www.nordicsemi.com/.
-	 */
-	if (ftpan_59()) {
-		NRF_MPU->DISABLEINDEBUG =
-			MPU_DISABLEINDEBUG_DISABLEINDEBUG_Disabled <<
-			MPU_DISABLEINDEBUG_DISABLEINDEBUG_Pos;
-	}
+	SystemInit();
```

Comment:
SystemInit() brings-in conditional code for implementing ftpan76, as well. That functionality was not present here; please, double-check whether there is a behavioral change.

Reply:
SystemCoreClock variable is defined globally (and set) in system_nrf51.c. We do not have to bring it back in soc.c file.

Reply:
No, we cannot bring it back because `uint32_t SystemCoreClock` is defined in "system_nrf51.c". We could call `SystemCoreClockUpdate()` right after `SystemInit()` but IMO it would be redundant. `SystemCoreClockUpdate()` is intended to update the `SystemCoreClock` according to clock register settings in a given CPU but `SystemInit()` is also supposed to do it (see [CMSIS template](https://github.com/ARM-software/CMSIS_5/blob/develop/Device/_Template_Vendor/Vendor/Device/Include/system_Device.h#L41)).
In our case for both nRF51 and nRF52 the clock speed is fixed, so the startup initialization of `SystemCoreClock` is sufficient (even the call to `SystemCoreClockUpdate()` at the end of `SystemInit()` for nRF52 is redundant).

Reply:
Regarding the PAN 76, the comment says that this is applies to nRF51802 only. I debugged it on my pca10028 and `is_peripheral_domain_setup_needed()` returns `false`.

Reply:
> > (even the call to SystemCoreClockUpdate() at the end of SystemInit() for nRF52 is redundant).

> Then it should be removed from nrf52, because this is source of confusion. the start-up code should be aligned as much as possible between SOCs

I agree. And that's a task for MDK guys. I'll let them know.